pyvrp/cpp/SubPopulation.cpp

```diff
@@ -0,0 +1,161 @@
+}
+
+std::vector<SubPopulation::Item>::const_iterator SubPopulation::cend() const
+{
+    return items.cend();
+}
+
+void SubPopulation::remove(
+    std::vector<SubPopulation::Item>::iterator const &iterator)
+{
+    for (auto &[individual, fitness, proximity] : items)
+        // Remove individual from other proximities.
+        for (size_t idx = 0; idx != proximity.size(); ++idx)
+            if (proximity[idx].second == iterator->individual)
+            {
+                proximity.erase(proximity.begin() + idx);
+                break;
+            }
+
+    delete iterator->individual;  // dispose of manually allocated memory
+    items.erase(iterator);        // before the item is removed.
+}
+
+void SubPopulation::purge()
+{
+    while (size() > params.minPopSize)
+    {
+        // Remove duplicates from the subpopulation (if they exist)
+        auto const pred = [&](auto &iterator)
+        {
+            return !iterator.proximity.empty()
+                   && *iterator.proximity[1].second == *iterator.individual;
+        };
+
+        auto const duplicate = std::find_if(items.begin(), items.end(), pred);
+
+        if (duplicate == items.end())  // there are no more duplicates
+            break;
+
+        remove(duplicate);
+    }
+
+    while (size() > params.minPopSize)
+    {
+        // Remove the worst individual (worst in terms of biased fitness)
+        updateFitness();
+
+        auto const worstFitness = std::max_element(
+            items.begin(),
+            items.end(),
+            [](auto const &a, auto const &b) { return a.fitness < b.fitness; });
+
+        remove(worstFitness);
+    }
+}
+
+void SubPopulation::updateFitness()
+{
+    if (items.empty())
+        return;
+
+    std::vector<size_t> byCost(size());
+    std::iota(byCost.begin(), byCost.end(), 0);
+
+    std::stable_sort(
+        byCost.begin(),
+        byCost.end(),
+        [&](size_t a, size_t b)
+        { return items[a].individual->cost() < items[b].individual->cost(); });
+
+    std::vector<std::pair<double, size_t>> diversity;
+    for (size_t rank = 0; rank != size(); rank++)
+    {
+        auto const dist = avgDistanceClosest(byCost[rank]);
+        diversity.emplace_back(dist, rank);
+    }
+
+    std::stable_sort(diversity.begin(), diversity.end(), std::greater<>());
+
+    auto const popSize = static_cast<double>(size());
+    auto const nbElite = std::min(params.nbElite, size());
+    auto const divWeight = 1 - nbElite / popSize;
+
+    for (size_t divRank = 0; divRank != size(); divRank++)
+    {
+        auto const costRank = diversity[divRank].second;
+        auto const idx = byCost[costRank];
+        items[idx].fitness = (costRank + divWeight * divRank) / popSize;
+    }
+}
+
+double SubPopulation::avgDistanceClosest(size_t idx) const
+{
+    auto const &item = items[idx];
+    auto const maxSize = std::min(item.proximity.size(), params.nbClose);
+
+    if (maxSize == 0)
+        return 0.0;  // TODO 0.0 or 1.0?
```

Comment:
I think 0.0 is fine, right? Either an item has no proximity, so it will always have the best diversity rank. Or when `nbClose=0`, then all individuals have the same diversity.